Issue #2325: Add snapshot and restore API for CircuitBreaker

kssumin opened the pull request:
## Problem

In production microservice environments, CircuitBreaker configurations often need to be updated at runtime - for example, adjusting failure thresholds based on observed system behavior or temporarily relaxing limits during known maintenance windows. However, the current API forces a painful choice: keep the old configuration with its accumulated operational data, or reconfigure and lose all circuit breaker history.

When reconfiguring a CircuitBreaker, creating a new instance discards:
- **Accumulated metrics** (success/failure counts, slow call statistics)
- **Current state** (OPEN, HALF_OPEN, etc.)
- **State-specific timing** (retry-after timestamps for OPEN state)
- **Attempts counter** (critical for HALF_OPEN progression)

This makes runtime reconfiguration impractical for production systems that need to preserve operational continuity. For instance, if a circuit breaker is in HALF_OPEN state testing service recovery, reconfiguring it would reset to CLOSED and lose all accumulated test attempts, potentially allowing a flood of requests to a still-unhealthy service.

**Example Impact**:
```java
// Circuit breaker has been accumulating metrics and is currently OPEN
CircuitBreaker cb = CircuitBreaker.of("payment-service", config);
// After many failures: state=OPEN, failureRate=75%, waitUntil=10:30:00

// Need to update configuration (e.g., increase threshold)
CircuitBreakerConfig newConfig = CircuitBreakerConfig.custom()
    .failureRateThreshold(60)  // Changed from 50
    .build();

// Currently, the only option loses all state
CircuitBreaker newCb = CircuitBreaker.of("payment-service", newConfig);
// Result: state=CLOSED, no metrics, retry-after timing lost
// Impact: Service might be overwhelmed by requests before OPEN state is re-established
```

This violates operational safety principles:
- Configuration changes shouldn't compromise circuit protection
- Operators cannot safely tune thresholds in production
- No way to persist/restore circuit breaker state for deployment scenarios

## Solution

The fix introduces a snapshot/restore mechanism that decouples state preservation from configuration. Operators can now capture a circuit breaker's operational state, apply new configuration, and restore the state - maintaining protection continuity during reconfiguration.

The approach uses an immutable snapshot pattern that captures:
- Circuit breaker state and the data needed to reconstruct it
- Aggregate metrics (counts, not sliding window internals)
- State-specific metadata (retry timing, attempt counters)

This enables safe hot-reconfiguration: the circuit breaker maintains its protective stance (e.g., stays OPEN if service is unhealthy) while adopting new threshold rules. The sliding window is intentionally reset with new configuration, as preserving window data across configuration changes would create semantic ambiguity (e.g., mixing data from different window sizes).

## Impact

**New Capabilities**:
- **Runtime reconfiguration** - Update thresholds/windows without losing protection state
- **State persistence** - Snapshot is Serializable for deployment continuity scenarios
- **Configuration migration** - Safely transition between different circuit breaker strategies

**API Changes**:
```java
// New methods on CircuitBreaker interface
CircuitBreakerSnapshot createSnapshot();

static CircuitBreaker of(String name,
                         CircuitBreakerConfig config,
                         CircuitBreakerSnapshot snapshot);
```

**Usage Pattern**:
```java
// Capture state before reconfiguration
CircuitBreakerSnapshot snapshot = circuitBreaker.createSnapshot();

// Apply new configuration while preserving operational state
CircuitBreaker reconfigured = CircuitBreaker.of(
    "payment-service",
    newConfig,
    snapshot
);
// Result: Same state and metrics, new configuration rules
```

**Compatibility**: This is a non-breaking addition to the API. Existing code continues to work unchanged. The feature enables new use cases previously impossible with the API.

Closes #2325

RobWin commented on resilience4j-circuitbreaker/src/test/java/io/github/resilience4j/circuitbreaker/CircuitBreakerSnapshotTest.java:
```diff
@@ -0,0 +1,290 @@
+        assertThat(cb2.getState())
+            .as("OPEN state should be preserved")
+            .isEqualTo(CircuitBreaker.State.OPEN);
+    }
```
@copilot open a new pull request to apply changes based on [this feedback](https://github.com/resilience4j/resilience4j/pull/2365#discussion_r2656366238)

kssumin replied:
Addressed in ef701bd9.

kssumin commented on resilience4j-circuitbreaker/src/main/java/io/github/resilience4j/circuitbreaker/internal/CircuitBreakerStateMachine.java:
```diff
@@ -463,6 +513,42 @@ private CircuitBreakerState getCircuitBreakerStateObjectFromState(State state){
+                    // Fallback if no retry time in snapshot
+                    yield new OpenState(snapshot.getAttempts(),
+                        CircuitBreakerMetrics.forClosed(circuitBreakerConfig));
+                }
```
I intentionally left this as-is because the existing state machine already handles this case naturally — when `tryAcquirePermission()` is called on an OPEN state, it checks whether the wait duration has elapsed and transitions to HALF_OPEN if so.

Adding explicit validation at restore time would mean the restore method makes state transition decisions, which breaks the separation of responsibility: **restore should faithfully reconstruct the snapshot state, and the state machine should handle transitions at runtime**.

From the caller's perspective, restoring an OPEN snapshot with a past `retryAfterWaitUntil` will still behave correctly — the very next `tryAcquirePermission()` call will trigger the transition to HALF_OPEN, which is the same behavior as a non-restored circuit breaker whose wait duration has already elapsed.

kssumin commented on resilience4j-circuitbreaker/src/test/java/io/github/resilience4j/circuitbreaker/CircuitBreakerSnapshotTest.java:
```diff
@@ -0,0 +1,290 @@
+        assertThat(metrics.getNumberOfSlowSuccessfulCalls()).isEqualTo(1);
+        assertThat(metrics.getNumberOfSlowFailedCalls()).isEqualTo(1);
+    }
+}
```
Addressed in 9230b658.

kssumin commented on resilience4j-circuitbreaker/src/main/java/io/github/resilience4j/circuitbreaker/CircuitBreakerSnapshot.java:
```diff
@@ -0,0 +1,386 @@
+                ", numberOfNotPermittedCalls=" + numberOfNotPermittedCalls +
+                '}';
+        }
+
```
Addressed in 3b236913.

kssumin commented on resilience4j-circuitbreaker/src/main/java/io/github/resilience4j/circuitbreaker/CircuitBreaker.java:
```diff
@@ -399,6 +399,44 @@ static CircuitBreaker of(String name,
+                             CircuitBreakerConfig config,
+                             CircuitBreakerSnapshot snapshot) {
+        return new CircuitBreakerStateMachine(name, config, snapshot);
+    }
```
Addressed in 7231b395.

kssumin commented on resilience4j-circuitbreaker/src/main/java/io/github/resilience4j/circuitbreaker/internal/CircuitBreakerStateMachine.java:
```diff
@@ -463,6 +513,42 @@ private CircuitBreakerState getCircuitBreakerStateObjectFromState(State state){
+                }
+            }
+            case HALF_OPEN -> new HalfOpenState(snapshot.getAttempts());
+            case FORCED_OPEN -> new ForcedOpenState(snapshot.getAttempts());
```
I intentionally left this as-is because `attempts=0` is a valid state for CLOSED, DISABLED, and METRICS_ONLY. Applying a blanket positive validation or silently normalizing to 1 would either reject valid snapshots or silently mutate data.

The primary consumer `createSnapshot()` always captures the actual attempts value from the current state, so invalid values can only come from manual builder construction — which is the caller's responsibility.

kssumin commented:
Thank you for the review, @RobWin. I've gone through all the Copilot findings and addressed them below.

### Addressed

| Feedback | Commit | Rationale |
|----------|--------|-----------|
| Null handling tests | 9230b658 | The new `of(name, config, snapshot)` and `CircuitBreakerSnapshot.Builder` are public APIs. Added 4 tests verifying `NullPointerException` contracts, matching existing patterns in `CircuitBreakerRegistryTest` and `CircuitBreakerTest`. |
| Automatic OPEN→HALF_OPEN transition test | 322b31a7 | A restored OPEN state must respect `retryAfterWaitUntil`. Added a `MockClock`-based test (same approach as `CircuitBreakerStateMachineTest`) verifying permission is denied before wait duration and HALF_OPEN transition occurs after. |
| Post-restore behavior tests | ef701bd9 | Restoring state is only meaningful if the circuit breaker continues to function afterward. Added 3 tests: new calls are tracked, failure threshold still triggers OPEN, and manual transition to HALF_OPEN works. |
| Tags preservation | 7231b395 | The original snapshot constructor always set `tags = emptyMap()`, while all other constructors accept tags. Added `of(name, config, snapshot, tags)` factory method and delegating constructor so tags are not silently lost. |
| MetricsBuilder negative validation | 11570b0b | `RetryConfig.Builder` and `BulkheadConfig.Builder` already validate inputs. Following this pattern, added non-negative checks to all 6 MetricsBuilder setters to catch invalid inputs early. |
| Serialization round-trip tests | f09d6513 | This PR positions `Serializable` as a key capability for state persistence. Added 2 tests (CLOSED and OPEN state) verifying all fields survive `ObjectOutputStream` → `ObjectInputStream` round-trip. |
| equals/hashCode | 3b236913 | `SnapshotImpl` in `resilience4j-core` already implements `equals/hashCode`. `CircuitBreakerSnapshot` is the same kind of immutable value object, so it follows the same convention. |
| Javadoc for factory method | 7231b395 | The snapshot-based `of()` creates a new instance with a fresh event processor. Documented that event subscriptions are not preserved and tags default to empty map unless the tags-aware overload is used. |

### Intentionally not addressed

I considered the remaining two suggestions carefully but decided not to apply them for the following reasons. Happy to revisit if you think otherwise.

| Feedback | Rationale |
|----------|-----------|
| `retryAfterWaitUntil` past-time validation | `OpenState.tryAcquirePermission()` already checks whether the wait duration has elapsed and transitions to HALF_OPEN. Adding validation at restore time would break responsibility separation — restore should faithfully reconstruct state, and the state machine should handle transitions at runtime. A past `retryAfterWaitUntil` simply triggers HALF_OPEN on the next `tryAcquirePermission()` call, same as a non-restored circuit breaker. |
| `attempts` positive validation | `attempts=0` is valid for CLOSED, DISABLED, and METRICS_ONLY states. Blanket positive validation would reject valid snapshots. Silently normalizing to 1 would mutate data without the caller's knowledge. `createSnapshot()` always captures the actual value, so invalid values can only come from manual builder construction — which is the caller's responsibility. |

Please let me know if there's anything else I should address.
